File: src/tray_status_icon.c
#include "tray_status_icon.h"

// GTK3-only includes - this file compiles separately with GTK3 flags
#include <gtk/gtk.h>
#include <gdk-pixbuf/gdk-pixbuf.h>

struct _TrayStatusIcon {
    GtkStatusIcon *status_icon;
    TrayStatusIconCallback callback;
    gpointer user_data;
};

static void on_tray_activate(GtkStatusIcon *status_icon, TrayStatusIcon *tray);
static void on_tray_popup_menu(GtkStatusIcon *status_icon, guint button, guint activate_time, TrayStatusIcon *tray);

TrayStatusIcon* tray_status_icon_new(void) {
    TrayStatusIcon *tray = g_malloc0(sizeof(TrayStatusIcon));
    
    // Create GTK3 status icon
    tray->status_icon = gtk_status_icon_new();
    
    // Set default properties
    gtk_status_icon_set_title(tray->status_icon, "Commodoro");
    gtk_status_icon_set_tooltip_text(tray->status_icon, "Commodoro Timer");
    gtk_status_icon_set_visible(tray->status_icon, TRUE);
    
    // Connect signals
    g_signal_connect(tray->status_icon, "activate", G_CALLBACK(on_tray_activate), tray);
    g_signal_connect(tray->status_icon, "popup-menu", G_CALLBACK(on_tray_popup_menu), tray);
    
    return tray;
}

void tray_status_icon_free(TrayStatusIcon *tray) {
    if (!tray) return;
    
    if (tray->status_icon) {
        g_object_unref(tray->status_icon);
    }
    
    g_free(tray);
}

void tray_status_icon_set_callback(TrayStatusIcon *tray, TrayStatusIconCallback callback, gpointer user_data) {
    if (!tray) return;
    
    tray->callback = callback;
    tray->user_data = user_data;
}

void tray_status_icon_update(TrayStatusIcon *tray, cairo_surface_t *surface, const char *tooltip) {
    if (!tray || !surface) return;
    
    // Convert Cairo surface to GdkPixbuf
    int width = cairo_image_surface_get_width(surface);
    int height = cairo_image_surface_get_height(surface);
    
    GdkPixbuf *pixbuf = gdk_pixbuf_get_from_surface(surface, 0, 0, width, height);
    if (pixbuf) {
        // Update icon
        gtk_status_icon_set_from_pixbuf(tray->status_icon, pixbuf);
        g_object_unref(pixbuf);
    }
    
    // Update tooltip
    if (tooltip) {
        gtk_status_icon_set_tooltip_text(tray->status_icon, tooltip);
    }
}

void tray_status_icon_set_visible(TrayStatusIcon *tray, gboolean visible) {
    if (!tray) return;
    
    gtk_status_icon_set_visible(tray->status_icon, visible);
}

gboolean tray_status_icon_is_embedded(TrayStatusIcon *tray) {
    if (!tray) return FALSE;
    
    return gtk_status_icon_is_embedded(tray->status_icon);
}

static void on_tray_activate(GtkStatusIcon *status_icon, TrayStatusIcon *tray) {
    (void)status_icon; // Suppress unused parameter warning
    
    if (tray->callback) {
        tray->callback("activate", tray->user_data);
    }
}

static void on_tray_popup_menu(GtkStatusIcon *status_icon, guint button, guint activate_time, TrayStatusIcon *tray) {
    (void)status_icon; // Suppress unused parameter warning
    (void)button;      // Suppress unused parameter warning
    (void)activate_time; // Suppress unused parameter warning
    
    if (tray->callback) {
        tray->callback("popup-menu", tray->user_data);
    }
}

File: src/gui/tray_status_icon.zig
const std = @import("std");
// Use workaround to avoid GdkEvent opaque type issues
const cw = @import("../c_workaround.zig");
const c = cw.c;

pub const TrayStatusIcon = struct {
    status_icon: *c.GtkStatusIcon,
    callback: ?*const fn (action: []const u8, user_data: ?*anyopaque) void,
    user_data: ?*anyopaque,
    allocator: std.mem.Allocator,
    
    const Self = @This();
    
    pub fn init(allocator: std.mem.Allocator) !*Self {
        const tray = try allocator.create(Self);
        
        // Create GTK3 status icon
        const status_icon = c.gtk_status_icon_new() orelse return error.StatusIconCreationFailed;
        
        tray.* = .{
            .allocator = allocator,
            .status_icon = status_icon,
            .callback = null,
            .user_data = null,
        };
        
        // Set default properties
        c.gtk_status_icon_set_title(status_icon, "Zigodoro");
        c.gtk_status_icon_set_tooltip_text(status_icon, "Zigodoro Timer");
        c.gtk_status_icon_set_from_icon_name(status_icon, "appointment-soon");
        c.gtk_status_icon_set_visible(status_icon, 1);
        
        // Connect signals
        _ = c.g_signal_connect_data(status_icon, "activate", @ptrCast(&onTrayActivate), tray, null, 0);
        _ = c.g_signal_connect_data(status_icon, "popup-menu", @ptrCast(&onTrayPopupMenu), tray, null, 0);
        
        return tray;
    }
    
    pub fn deinit(self: *Self) void {
        c.g_object_unref(self.status_icon);
        self.allocator.destroy(self);
    }
    
    pub fn setCallback(self: *Self, callback: *const fn ([]const u8, ?*anyopaque) void, user_data: ?*anyopaque) void {
        self.callback = callback;
        self.user_data = user_data;
    }
    
    pub fn update(self: *Self, surface: ?*anyopaque, tooltip: []const u8) void {
        if (surface) |surf| {
            // Update the status icon with the Cairo surface
            const width = c.cairo_image_surface_get_width(surf);
            const height = c.cairo_image_surface_get_height(surf);
            
            std.log.info("Updating status icon with Cairo surface: {}x{}", .{width, height});
            
            const pixbuf = c.gdk_pixbuf_get_from_surface(surf, 0, 0, width, height);
            if (pixbuf) |pb| {
                c.gtk_status_icon_set_from_pixbuf(self.status_icon, pb);
                c.g_object_unref(pb);
            } else {
                std.log.err("Failed to create pixbuf from Cairo surface", .{});
            }
        } else {
            std.log.warn("No Cairo surface provided to status icon update", .{});
        }
        self.updateTooltip(tooltip);
    }
    
    pub fn updateTooltip(self: *Self, tooltip: []const u8) void {
        var tooltip_z: [256]u8 = undefined;
        const tooltip_str = std.fmt.bufPrintZ(&tooltip_z, "{s}", .{tooltip}) catch return;
        c.gtk_status_icon_set_tooltip_text(self.status_icon, tooltip_str);
    }
    
    pub fn setVisible(self: *Self, visible: bool) void {
        c.gtk_status_icon_set_visible(self.status_icon, if (visible) @as(c.gboolean, 1) else @as(c.gboolean, 0));
    }
    
    pub fn isEmbedded(self: *const Self) bool {
        return c.gtk_status_icon_is_embedded(self.status_icon) == @as(c.gboolean, 1);
    }
    
    fn onTrayActivate(status_icon: *c.GtkStatusIcon, tray: *Self) callconv(.C) void {
        _ = status_icon;
        
        if (tray.callback) |cb| {
            cb("activate", tray.user_data);
        }
    }
    
    fn onTrayPopupMenu(status_icon: *c.GtkStatusIcon, button: c.guint, activate_time: c.guint, tray: *Self) callconv(.C) void {
        _ = status_icon;
        _ = button;
        _ = activate_time;
        
        if (tray.callback) |cb| {
            cb("popup-menu", tray.user_data);
        }
    }
};